### PR TITLE
Migrating remaining JS modules to Typescript (part 1)

### PR DESCRIPTION
**Pull request by smallsaucepan**
Migrating remaining Javascript modules to Typescript (the main implementation file at least). More discussion in #2538 

This PR covers:

turf-difference
turf-dissolve
turf-ellipse
turf-envelope
turf-explode
turf-flatten
turf-flip
turf-point-on-feature
turf-points-within-polygon
turf-polygon-smooth
turf-polygon-tangents
turf-rewind
turf-sample
turf-sector
turf-shortest-path
turf-simplify
turf-tag

Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Have read [How To Contribute](https://github.com/Turfjs/turf/blob/master/CONTRIBUTING.md#how-to-contribute).
- [x] Run `npm test` at the sub modules where changes have occurred.
- [x] Run `npm run lint` to ensure code style at the turf module level.

### CODE REVIEW

File: packages/turf-rewind/index.ts
Comment (mfedderly) [inline packages/turf-rewind/index.ts:50]: Possibly a little sketchy but we can skip this whole copy for Point and MultiPoint geometries. Or maybe just don't even accept them into the function in the first place.

File: packages/turf-sample/index.ts
Comment (mfedderly) [inline packages/turf-sample/index.ts:26]: JS error handling

File: packages/turf-shortest-path/index.ts
Comment (mfedderly) [inline packages/turf-shortest-path/index.ts:56]: I'm kind of surprised that this doesn't also allow MultiPolygons from the JSDoc above where it accepts `Feature`, but this matches the previous types from the deleted .d.ts.

File: packages/turf-shortest-path/index.ts
Comment (mfedderly) [inline packages/turf-shortest-path/index.ts:70]: NIce bugfix swapping the `()` around

File: packages/turf-shortest-path/lib/javascript-astar.d.ts
Comment (mfedderly) [inline packages/turf-shortest-path/lib/javascript-astar.d.ts:1]: We can probably just update definitely-typed directly, that's usually pretty quick to do.
Otherwise we probably need to preserve whatever license blob is relevant from upstream.

File: packages/turf-simplify/lib/simplify.d.ts
Comment (mfedderly) [inline packages/turf-simplify/lib/simplify.d.ts:1]: Can we just depend on simplify-js directly and remove our local copy?
Otherwise we probably need to preserve license text here as well.

File: packages/turf-dissolve/index.ts
Comment (smallsaucepan) [reply packages/turf-dissolve/index.ts:51]: Have updated it to ```[Geom, ...Geom[]]```. The interface to union() expects an initial Geom, then any extra Geoms are spread in as rest parameters (rather than an array).

File: packages/turf-dissolve/index.ts
Comment (smallsaucepan) [reply packages/turf-dissolve/index.ts:54]: You're correct. Have removed.

File: packages/turf-sample/index.ts
Comment (smallsaucepan) [reply packages/turf-sample/index.ts:26]: Reinstated!

File: packages/turf-rewind/index.ts
Comment (smallsaucepan) [reply packages/turf-rewind/index.ts:47]: Fixed. Searched for pattern in other files as well and fixed those too.

File: packages/turf-rewind/index.ts
Comment (smallsaucepan) [reply packages/turf-rewind/index.ts:36]: Agreed. Reinstated.

File: packages/turf-rewind/index.ts
Comment (smallsaucepan) [reply packages/turf-rewind/index.ts:50]: Have added some logic to only clone if not Point or MultiPoint.

File: packages/turf-points-within-polygon/index.ts
Comment (smallsaucepan) [reply packages/turf-points-within-polygon/index.ts:65]: Agree - logic reinstated.

File: packages/turf-flip/index.ts
Comment (smallsaucepan) [reply packages/turf-flip/index.ts:30]: Done. And replaced in other areas where misuse appeared.

File: packages/turf-shortest-path/index.ts
Comment (smallsaucepan) [reply packages/turf-shortest-path/index.ts:56]: Yeah, I haven't pulled too much at loose threads like this 😅

File: packages/turf-flatten/index.ts
Comment (smallsaucepan) [reply packages/turf-flatten/index.ts:47]: Personally I find the nested ternaries pretty difficult to parse, so the overloaded function definitions are more readable. Possibly also what the casual consumer would prefer?

File: packages/turf-flip/index.ts
Comment (smallsaucepan) [reply packages/turf-flip/index.ts:30]: @mfedderly do we also want to do this with the options object? E.g. a pattern like so:

```typescript
options?: {
  thisProp?: string,
  thatProp?: boolean,
}

// options could be anything e.g. undefined, empty object, string, etc
options = options ?? {} // defined but user could have still passed a string ...
if (!isObject(options)) throw new Error("options not an object"); // def got an object by now
const thisProp = options.thisProp ?? "some default"
const thatProp = options.thatProp ?? false;
```

File: packages/turf-shortest-path/lib/javascript-astar.d.ts
Comment (smallsaucepan) [reply packages/turf-shortest-path/lib/javascript-astar.d.ts:1]: If you're happy to, I'll add a license blob as a temporary measure so this PR can proceed. In the background will see what can be done to fix the types at Definitely Typed and tidy up down the track.

File: packages/turf-flip/index.ts
Comment (mfedderly) [reply packages/turf-flip/index.ts:30]: Yeah I was kind of thinking about this. Perhaps we want something like this everywhere and then pull args from DefaultFooOptions as we need it.

```
export interface FooOptions {
   ...
}
const DefaultFooOptions: FooOptions = {};
export default (options: Partial<FooOptions>) => {}
```

File: packages/turf-flatten/index.ts
Comment (mfedderly) [reply packages/turf-flatten/index.ts:47]: Yeah I maybe buy that argument. The ternaries aren't too bad if they're formatted appropriately it winds up like a big if/else, but its maybe harder to know how to read them as a consumer if you're getting some inferred type instead of something concrete.

File: packages/turf-rewind/index.ts
Comment (mfedderly) [reply packages/turf-rewind/index.ts:50]: The thing that's sketchy is that if you use rewind assuming that its then safe to mutate the return because the rewind cloned it, you suddenly have different behavior on Point and MultiPoint which is confusing. Sorry we maybe should just always clone.

File: packages/turf-shortest-path/lib/javascript-astar.d.ts
Comment (mfedderly) [reply packages/turf-shortest-path/lib/javascript-astar.d.ts:1]: I'm not a lawyer but that seems fine. Also this being A* explains why the pathing looks weird, I bet Dijkstra's would be slower but give a more intuitive looking answer.

**Review by mfedderly — APPROVED** [review]
looks like the force push hid all the changes from the github review window for me, but I'm going to just approve to unblock since we're definitely working different hours at the moment. Please clean up the license stuff before merging though.

I think we've been using `/*!` to denote license comments
https://github.com/Turfjs/turf/blob/master/packages/turf-great-circle/lib/arc.js#L1

**Comment by smallsaucepan** [comment]
Whoops. Sorry Matt. To summarise, agreed with pretty much everything you pointed out and will address with another commit shortly.

Only things I'm not sure I could address are some overloads and local lib typings. For overloads will leave as is as they seem more readable to the casual user. For local lib typings will put license in as a temporary measure and if there's something published (or that we can update easily) on Definitely Typed will tidy up in a subsequent PR.

Thanks for approving!

File: packages/turf-simplify/lib/simplify.d.ts
Comment (smallsaucepan) [reply packages/turf-simplify/lib/simplify.d.ts:1]: Included license text for now.

File: packages/turf-polygon-smooth/index.ts
Comment (smallsaucepan) [reply packages/turf-polygon-smooth/index.ts:45]: Realised I was being too restrictive with the types. The implementation is totally fine with a mixture of both Polygon and MultiPolygon. Have removed the overload and now the one signature accepts all permutations.

File: packages/turf-shortest-path/lib/javascript-astar.d.ts
Comment (smallsaucepan) [reply packages/turf-shortest-path/lib/javascript-astar.d.ts:1]: Included the license text as a workaround for now.

File: packages/turf-polygon-tangents/index.ts
Comment (smallsaucepan) [reply packages/turf-polygon-tangents/index.ts:114]: Yep, that should be fine. enext looked like it was being reused between the main function and a sub function, though only as a local var in the sub function.

**Comment by markcarroll** [comment]
@smallsaucepan can we get a version bump of the 7.0 alpha so we can test these out?

**Pull request merged**
